### ISSUE TITLE
Idea: Suggest Assert.Multiple in tests where Assert statements are called back to back

### ISSUE THREAD

**Issue by SeanKilleen** [nunit/nunit.analyzers#389]
Just happened to be thinking -- a lot of folks don't know about [NUnit Multiple Asserts](https://docs.nunit.org/articles/nunit/writing-tests/assertions/multiple-asserts.html) and how it can be helpful in situations where you may want multiple distinct assertions that don't rise to the level of separate tests. 

In tests that contain multiple asserts -- or where asserts exist one after another -- I think it'd be helpful to suggest this.

**Comment by manfred-brands** [comment]
It would depend on whether those test are independent or not.
Not for:
```
Assert.That(instance, Is.Not.Null);
Assert.That(instance.Property, ...);
```
Yes for:
```
Assert.That(instance.Property1, ...);
Assert.That(instance.Property2, ...);
Assert.That(instance.Property3, ...);
```

How to determine that is the challenge.

**Comment by mikkelbu** [comment]
@manfred-brands I was thinking in the same lines. We often have test like in the first case
```csharp
            var result = SomeCalculationThatReturnsSomeKindOfOption/Maybe();

            Assert.True(result.IsSuccess);
            Assert.Equal(expectedValue, result.Value.Value);
```

I don't have that much experience with Assert.Multiple as we don't use NUnit at work, but I think the best approach would be to find good real life candidates of "sequences of asserts" were it would make sense to convert them to Assert.Multiple and then see if we can extract some "conservative" pattern from this (by conservative I mean something that will only catch the second case).

**Comment by SeanKilleen** [comment]
@manfred-brands I guess it's a matter of opinion. The way I teach test automation, the example you provide:

```csharp
Assert.That(instance, Is.Not.Null);
Assert.That(instance.Property, ...);
```

Should be represented as two separate tests, because you want to have the additional triangulation if something should not be operating as you'd expect. 

However, if the tests _are_ related, we run into a problem -- what happens if both assertions will fail? If I fix the first assertion, the test will still fail because it hits the second assertion. Better to understand that both of those assertions, combined, are the purpose of the test. So I am of the opinion that tests should have one "logical" assertion, and if that assertion takes the form of multiple `Assert` statements, then those are a good candidate for `Asssert.Multiple`.

With that said, I'm all for starting conservative as @mikkelbu suggests. 

Perhaps the rule could be re-phrased as "Multiple asserts involving the same object". This would catch the most egregious case of multiple assertions against an object's properties, which is also where `Assert.Multiple` could be very useful.

**Comment by manfred-brands** [comment]
I agree that would be a good candidate for an initial version.

**Comment by SeanKilleen** [comment]
Nice! 🎉
